crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -359,12 +359,13 @@ impl<'db> SemanticIndexBuilder<'db> {
         self.visit_expr(&generator.iter);
         self.push_scope(scope);
 
-        self.current_assignment = Some(CurrentAssignment::Comprehension {
-            node: generator,
-            first: true,
-        });
+        self.current_assignments
+            .push(CurrentAssignment::Comprehension {
+                node: generator,
+                first: true,
+            });
         self.visit_expr(&generator.target);
-        self.current_assignment = None;
+        debug_assert!(self.current_assignments.pop().is_some());
```

Comment:
You want to move this out of a debug assertion or the `pop` call only happens during debugging. It probably makes sense to move this into a helper function. 

```suggestion
				let assignment = self.current_assignments.pop();
				debug_assert_ne!(assignment, None);
```

Reply:
right, of course. Thanks for pointing that out.

Reply:
Went with adding `push_assignment` and `pop_assignment` as a pair of functions for this operation.

I thought about `with_assignment` a bit but I am unsure of the relative cost of closure building here. Given the final assert at the end of building being in place, might not be much of an issue in practice.